auth_saml/models/res_users.py

```diff
@@ -0,0 +1,203 @@
+        and the interesting code is inside the "except" clause.
+        """
+        try:
+            # Attempt a regular login (via other auth addons) first.
+            return super()._check_credentials(password)
+
+        except (AccessDenied, passlib.exc.PasswordSizeError):
+            # since normal auth did not succeed we now try to find if the user
+            # has an active token attached to his uid
+            token = (
+                self.env["auth_saml.token"]
+                .sudo()
+                .search(
+                    [
+                        ("user_id", "=", self.env.user.id),
+                        ("saml_access_token", "=", password),
+                    ]
+                )
+            )
+            if not token:
+                raise AccessDenied()
+
+    def _autoremove_password_if_saml(self):
+        """Helper to remove password if it is forbidden for SAML users."""
+        if self.env.context.get("auth_saml_no_autoremove_password"):
+            return
+        if self._allow_saml_and_password():
+            return
+        to_remove_password = self.filtered(
+            lambda rec: rec.id != SUPERUSER_ID and rec.saml_ids and rec.password
+        )
+        if to_remove_password:
+
+            to_remove_password.with_context(
+                auth_saml_no_autoremove_password=True
+            ).write({"password": self._autoremove_password_if_saml_gen_password()})
+
+    @api.model
+    def _autoremove_password_if_saml_gen_password(self):
+        """
+        If password_security is installed, we cannot have a False-y password.
+        This is to avoid a very small bridging/compatbility module.
+        """
+        new_password = False
+        modules = self.env["ir.module.module"].sudo()._installed().keys()
+        if "password_security" in modules:
+            new_password = passlib.utils.generate_password(size=24, charset="ascii_72")
+        return new_password
+
+    def write(self, vals):
+        result = super().write(vals)
+        self._autoremove_password_if_saml()
+        self._ensure_saml_token_exists()
+        return result
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        result = super().create(vals_list)
+        result._autoremove_password_if_saml()
+        result._ensure_saml_token_exists()
+        return result
+
+    @api.model
+    def _allow_saml_and_password(self):
+        """Can both SAML and local password auth methods can coexist."""
+        return tools.str2bool(
+            self.env["ir.config_parameter"]
+            .sudo()
+            .get_param("auth_saml.allow_saml_uid_and_internal_password", "True")
+        )
+
+    def _ensure_saml_token_exists(self):
+        # workaround for Odoo not seeing the newly created auth_saml.token on
+        # the very first login
+        model_token = self.env["auth_saml.token"].sudo()
+        for record in self.filtered(lambda r: r.saml_ids):
```

Comment:
This also causes a problem when trying to change the language for example. It triggers the write method which triggers this one and you have an access error from a regular user.

Reply:
I'm having issues reproducing/understanding this one Paul.. Is it just a case of logging in as a user and changing the language? Or?

Reply:
Yes, just log in as a regular user with no administrative privileges and try to change the language.

Reply:
It comes from the fact that the user itself is writing on the res.user object (lang_id change) which triggers the write that triggers this method and tries to read own_user.saml_ids. This winds up in an access error because he does not have permissions to read the saml_ids even though they are attached to his own user